ros2cli/ros2cli/daemon/__init__.py

```diff
@@ -63,74 +64,91 @@ def make_xmlrpc_server():
-                node.get_subscriptions_info_by_topic,
-                node.get_service_names_and_types_by_node,
-                node.get_client_names_and_types_by_node,
-                bind(rclpy.action.get_action_server_names_and_types_by_node, node),
-                bind(rclpy.action.get_action_client_names_and_types_by_node, node),
-                node.count_publishers,
-                node.count_subscribers,
-                node.count_clients,
-                node.count_services
-            ]
-
-            server.register_introspection_functions()
-            for func in functions:
-                server.register_function(
-                    before_invocation(
-                        func, pretty_print_call))
-
-            shutdown = False
-
-            # shutdown the daemon in case of a timeout
-            def timeout_handler():
-                nonlocal shutdown
+    ros_domain_id = get_ros_domain_id()
+    node_args = argparse.Namespace(
+        node_name_suffix=f'_daemon_{ros_domain_id}_{uuid.uuid4().hex}',
+        start_parameter_services=False,
+        start_type_description_service=False)
+    with NetworkAwareNode(node_args) as node:
+        functions = [
+            node.get_name,
+            node.get_namespace,
+            node.get_node_names_and_namespaces,
+            node.get_node_names_and_namespaces_with_enclaves,
+            node.get_topic_names_and_types,
+            node.get_service_names_and_types,
+            bind(rclpy.action.get_action_names_and_types, node),
+            node.get_publisher_names_and_types_by_node,
+            node.get_publishers_info_by_topic,
+            node.get_subscriber_names_and_types_by_node,
+            node.get_subscriptions_info_by_topic,
+            node.get_service_names_and_types_by_node,
+            node.get_client_names_and_types_by_node,
+            bind(rclpy.action.get_action_server_names_and_types_by_node, node),
+            bind(rclpy.action.get_action_client_names_and_types_by_node, node),
+            node.count_publishers,
+            node.count_subscribers,
+            node.count_clients,
+            node.count_services
+        ]
+
+        # Dealing with the timeouts in this server is a bit tricky.  The caller
+        # passes in an overall inactivity timeout via the 'timeout' parameter;
+        # this server should quit when there is no activity within that timeout.
+        # The 'server.timeout' specifies how long 'server.handle_request()'
+        # should wait in 'select()' before returning with no work to do.  We set
+        # the 'server.timeout' to 200 milliseconds so we will react fairly
+        # quickly to external signals and quit.  To deal with the overall
+        # timeout, each function that is called resets the timer.  The result
+        # of all of this is that we properly react to signals (SIGINT and
+        # SIGTERM), we properly timeout when we are idle, and we properly quit
```

Comment:
this PR works with `SIGTERM`.

i do not quite figure out why `SIGINT` can terminate the server right away without this fix but `SIGTERM`, do you have any idea? cz if we can make this work in `SIGINT`, we do not need to poll with 200 msec.

Reply:
It's a good question.  I don't have an exact answer for you, but I do have some idea.

Take a look at https://gist.github.com/clalancette/0b55a26e71708089013377711b110593 .  That is a simplified version of what is going on in the daemon.  If you take that code and run it as-is, you can kill it with both SIGTERM and SIGINT.

If you uncomment https://gist.github.com/clalancette/0b55a26e71708089013377711b110593#file-selector-py-L53 (the call to `rclpy.init`) and run it, then SIGTERM no longer works.  I know that rclpy installs its own signal handling during initialization, in https://github.com/ros2/rclpy/blob/rolling/rclpy/src/rclpy/signal_handler.cpp .  Theoretically, that signal handling is supposed to call the "default" signal handling first, but it seems like possibly that is not working as intended.  I'm not sure.

Reply:
yeah i think so too, but i am not sure why yet either.

what do we want to do with this PR? it seems that the problem could be `rclpy` signal handler, but not sure at this moment. this PR actually fixes the problem, but problem is not so severe for user experience? so we should wait and check the signal handling in rclpy1st?

i feel like we can take this since this includes good enhancements, and after we can find the bug in rclpy, we can just adjust the timeout for the socket? what do you think? @clalancette @cottsay

Reply:
OK, a bit more information.

We install the [sigint](https://github.com/ros2/rclpy/blob/cbc21802aaada05c50541067a1cc84bcc0cd48a1/rclpy/src/rclpy/signal_handler.cpp#L359-L360) and [sigterm](https://github.com/ros2/rclpy/blob/cbc21802aaada05c50541067a1cc84bcc0cd48a1/rclpy/src/rclpy/signal_handler.cpp#L389-L390) signal handlers during `rclpy.init`.  In both cases, we save the original signal handler that was installed prior to us installing our own (that's the `g_original_sigint_handler` and `g_original_sigterm_handler`).

Later on, when we are handling signals, we call out to these original handlers, only if they are not NULL.  In the case of `g_original_sigint_handler`, the original handler function is not NULL, so it is called.  However, the `g_original_sigterm_handler` *is* NULL, so there is nothing for us to call.  That actually holds for a plain C program as well (you can see that in https://gist.github.com/clalancette/46487a6d6d447173a5bafc28068fe7e2).

It is not at all clear to me what to do here.  Given that the default action for SIGTERM is TERMINATE, `rclpy` could always do that in the absence of another signal handler being installed.  But that would be assuming that nobody changed SIGTERM to IGNORE, or something like that.

Reply:
> i feel like we can take this since this includes good enhancements, and after we can find the bug in rclpy, we can just adjust the timeout for the socket? what do you think? @clalancette @cottsay

Yeah, I personally think we should take this PR.  It should improve the user experience a bit, and doesn't have too much downside.  We can always revisit it if we decide to take a different path with rclpy signal handling.

Reply:
> However, the `g_original_sigterm_handler` _is_ NULL, so there is nothing for us to call. That actually holds for a plain C program as well...

This kinda glosses over that fact that a completely unhandled signal will cause the program to quit, so if we were really trying to be make the rclpy signal handling transparent, we'd exit the program on the spot if `g_original_sigterm_handler` is NULL when SIGTERM is received.

Reply:
> Ah, I guess this is what you're saying in the next paragraph, nevermind.

Yeah.  The more I look at it, the more it seems like a deliberate decision in `rclpy` to handle SIGTERM this way.  It makes *some* sense; if we were just to quit, then we would never clean up and there would be no reason to install a custom signal handler.  But it is kind of odd, in that it behaves differently from "normal" Unix programs.